book-content/chapters/08-classes.md

```diff
@@ -0,0 +1,1044 @@
+
+#### You Don't Always Need To Type Class Properties
+
+As we'll see, TypeScript can do some really smart inference with classes. It's able to infer the types of the properties from where we assign them in the constructor, so we can actually drop some of the type annotations:
+
+```typescript
+class Album {
+  title;
+  artist;
+  releaseYear;
+
+  constructor() {
+    this.title = "Loop Finding Jazz Records";
+    this.artist = "Jan Jelinek";
+    this.releaseYear = 2001;
+  }
+}
+```
+
+However, it's common to see the types specified in the class body as well since they act as a form of documentation for the class that's quick to read.
+
+### Adding Arguments to the Constructor
+
+We can use the constructor to declare arguments for the class. This allows us to pass in values when creating a new instance of the class.
+
+Update the constructor to accept an `opts` argument that includes the properties of the `Album` class:
+
+```typescript
+// inside the Album class
+constructor(opts: { title: string; artist: string; releaseYear: number }) {
+ // ...
+}
+```
+
+Then inside of the body of the constructor, we'll use assign `this.title`, `this.artist`, and `this.releaseYear` to the values of the `opts` argument.
+
+```typescript
+// inside the Album class
+constructor(opts: { title: string; artist: string; releaseYear: number }) {
+  this.title = opts.title;
+  this.artist = opts.artist;
+  this.releaseYear = opts.releaseYear;
+}
+```
+
+The `this` keyword refers to the instance of the class, and it's used to access the properties and methods of the class.
+
+Now, when we create a new instance of the `Album` class, we can pass an object with the properties we want to set. If we don't provide any values, the default values will be used:
+
+```typescript
+const loopFindingJazzRecords = new Album({
+  title: "Loop Finding Jazz Records",
+  artist: "Jan Jelinek",
+  releaseYear: 2001,
+});
+
+console.log(loopFindingJazzRecords.title); // Output: Loop Finding Jazz Records
+
+const unknownAlbum = new Album();
+
+console.log(unknownAlbum.title); // Output: Unknown Album
+```
+
+### Using a Class as a Type
+
+An interesting property of classes in TypeScript is that they can be used as types for variables and function parameters. The syntax is similar to how you would use any other type or interface.
+
+In this case, we'll use the `Album` class to type the `album` parameter of a `printAlbumInfo` function:
+
+```typescript
+function printAlbumInfo(album: Album) {
+  console.log(
+    `${album.title} by ${album.artist}, released in ${album.releaseYear}.`,
+  );
+}
+```
+
+We can then call the function and pass in an instance of the `Album` class:
+
+```typescript
+printAlbumInfo(sixtyNineLoveSongsAlbum);
+
+// Output: 69 Love Songs by The Magnetic Fields, released in 1999.
+```
+
+While using a class as a type is possible, it's a much more common pattern to require classes to implement a specific interface.
```

Comment:
this feels pretty opinionated, in many codebases I've seen classes were used as types directly - if you don't share an interface between multiple classes then why would I bother redeclaring every property/method to get that interface?